### ISSUE TITLE
Cannot get linear speedup on faster rcnn with multi-gpu&windows

### ISSUE THREAD

**Issue by YuwenXiong** [apache/incubator-mxnet#5237]
## Environment info
Operating System: Windows Server 2012 R2

Compiler: VS2015

Package used (Python/R/Scala/Julia): Python

MXNet version: Both tried latest prebuild binnary from https://github.com/yajiedesign/mxnet/releases and installed from source with latest commit.

MXNet commit hash (`git rev-parse HEAD`): 95d9b59

Python version and distribution: Anaconda 2, python 2.7.12

## Steps to reproduce

1. Run python train_end2end.py --network resnet --gpus 0
2. Run python train_end2end.py --network resnet --gpus 0,1
3. Run python train_end2end.py --network resnet --gpus 0,1,2,3
4. The first one will have 2.2 images/sec, but the second one will only have 3.0 images/sec, the third one will only have 3.5 images/sec (Pascal TTX speed), tested on Pascal TTX and TTX. CPU are 2 E5 2650-v3, and CPU usage will be around 10% when running.

**Comment by tornadomeet** [comment]
this should be normal at current implementation, the io part and some operator is written with python which is not optimized much.

**Comment by Orpine** [comment]
Hi @tornadomeet, I tested it on a Linux server with 4 K40 GPU, it can achieve linear speedup, and I used the CXX Proposal operator.

**Comment by tornadomeet** [comment]
i implement `train_end2end.py` and test in a linux server with 6 k80 at that time, it will not get linear spped up. so have you rewritten the proposal operator?

**Comment by Orpine** [comment]
I take C++ operator from https://github.com/precedenceguo/mx-rcnn/tree/master/operator. I think the python operator should not be the critical point.

**Comment by Orpine** [comment]
It seems to be a IO problem, when I skip the get_batch function and use use a pre-computed mini-batch it could get linear speedup. But it supposed to be async and won't have such huge impact, right? And it works well on Linux, so I wonder what's wrong with it, does windows has any limitation?

**Comment by tornadomeet** [comment]
i'm not familiar with windows about the python io part , cc @yajiedesign

**Comment by yajiedesign** [comment]
I am seems it cpu problem . get_rcnn_batch not async and it more cpu calculation, and it is run in single thread.
You can change it into a multithreaded mode,
As to why linux linear speedup.I think it's because K40 isn't fast enough.

**Comment by Orpine** [comment]
@yajiedesign Many thanks! I think you are right, I tested Linux and Windows on the same machine with 2 TTX, the problem both exists on two OS. Could you please tell me how to `change it into a multithreaded mode`? I found that when I use two cards, the get_batch() function time cost is also doubled, cc @precedenceguo

**Comment by niluanwudidadi** [comment]
Hi， @Orpine when I use mx-rcnn in windows，I have met a problem，after I successfully installed mxnet I can‘t compile cython，so what do you do in this part，thank for your help。

**Comment by Orpine** [comment]
@niluanwudidadi  please refer this repo: https://github.com/li-haoran/mx-rcnn, it contains setup.py for windows.

**Comment by niluanwudidadi** [comment]
Hello,again me. @Orpine I have met a problem in multi-gpu rcnn training,when i use 4 1080, the speed is no faster than 2 gpu, and the terminal report the message "only 4 out of 12 GPU pairs are enabled direct access.The command is python train_end2end --gpu 0,1,2,3. Is the vgg network's reason?when I use resnet the speed is increases linearly with the number of gpu; and is there some parameters should set?

**Comment by precedenceguo** [comment]
Indeed, get_batch is single threaded. I think yajie means we should move the io to something like ImageRecordIO.

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to ping me to reopen if this is still an active issue. Thanks!

**Comment by 315386775** [comment]
@precedenceguo @YuwenXiong how to change get_batch into a multithreaded mode? Thanks
